Signaling Game Experiments

vbhatt-cs opened the pull request:
I have added the code for running tabular Q-Learning and DQN on signaling game. I also modified tabular_qlearner.py a bit to allow centralized training, but it currently only uses joint observation and not joint action. 

There is a lot of common code between `lewis_signaling_dqn.py` and `lewis_signaling_qlearner.py` (mainly the training loop and plotting code). I can move these to a different place if necessary. I could also merge them into a single file.

ElnazDavoodi commented on open_spiel/games/lewis_signaling.cc:
```diff
@@ -135,12 +135,16 @@ void LewisSignalingState::ObservationTensor(Player player,
-    (*values)[offset + message_] = 1;
-    offset += num_messages_;
+  if (static_cast<Players>(player) == Players::kSender) {
+    if (state_ != -1) {
```
only in the last step state_ == -1, is that right?

ElnazDavoodi replied:
only in the last step state_ == -1, is that right?

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+                    "Payoffs to use ('random' for random [0, 1) payoffs)")
+
+# Alg parameters
+flags.DEFINE_bool("centralized", False, "Set to use centralized Q-Learning")
```
Remove it, the flag is not used in dqn

ElnazDavoodi replied:
Remove it, the flag is not used in dqn

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+    env = rl_environment.Environment(game, **env_configs)
+    state_size = env.observation_spec()["info_state"][0]
+    num_actions = env.action_spec()["num_actions"]
+    replay_buffer_capacity = int(1e4)
```
Add int(1e4) as Flag.

ElnazDavoodi replied:
Add int(1e4) as Flag.

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+            for cur_episode in range(training_episodes):
+                time_step = env.reset()
+                # Find cur_state for logging. See lewis_signaling.cc for info_state details
+                cur_state = time_step.observations["info_state"][0][3:].index(1)
```
how about adding current state to time_step.observation["state"]?

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+            base_info_state1 = [0.0, 1.0, 0.0] + [0.0] * num_states
+
+            for s in range(num_states):
+                info_state0 = copy.deepcopy(base_info_state0)
```
Not sure if you need a copy! why not passig base_info_state0 to agents[0]._epsilon_greedy?

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+
+            for s in range(num_states):
+                info_state0 = copy.deepcopy(base_info_state0)
+                info_state0[3 + s] = 1.0
```
use a value specified as a Flag instead of 3

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+                m, _ = agents[0]._epsilon_greedy(info_state0,
+                                                 np.arange(num_messages), 0)
+                info_state1 = copy.deepcopy(base_info_state1)
+                info_state1[3 + m] = 1.0
```
use a value specified as a Flag instead of 3

ElnazDavoodi replied:
use a value specified as a Flag instead of 3

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+                info_state0[3 + s] = 1.0
+                m, _ = agents[0]._epsilon_greedy(info_state0,
+                                                 np.arange(num_messages), 0)
+                info_state1 = copy.deepcopy(base_info_state1)
```
Not sure if you need a copy! why not passing base_info_state1 to agents[1]._epsilon_greedy?

ElnazDavoodi replied:
Not sure if you need a copy! why not passing base_info_state1 to agents[1]._epsilon_greedy?

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+            if scalar_labels is None:
+                scalar_labels = [None] * len(scalars)
+
+            assert len(scalars) == len(
```
use assertEqual

ElnazDavoodi replied:
use assertEqual

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_qlearner.py:
```diff
@@ -0,0 +1,297 @@
+        for cur_episode in range(training_episodes):
+            time_step = env.reset()
+            # Find cur_state for logging. See lewis_signaling.cc for info_state details
+            cur_state = time_step.observations["info_state"][0][3:].index(1)
```
maybe add current state to time_step.observations["state"]?

ElnazDavoodi replied:
maybe add current state to time_step.observations["state"]?

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_qlearner.py:
```diff
@@ -0,0 +1,297 @@
+            base_info_state1 = [base_info_state1, base_info_state1.copy()]
+
+        for s in range(num_states):
+            info_state0 = copy.deepcopy(base_info_state0)
```
do you need a copy of base_info_state0 to pass it to agents[0]._epsilon_greedy?

ElnazDavoodi replied:
do you need a copy of base_info_state0 to pass it to agents[0]._epsilon_greedy?

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_qlearner.py:
```diff
@@ -0,0 +1,297 @@
+        for s in range(num_states):
+            info_state0 = copy.deepcopy(base_info_state0)
+            if centralized:
+                info_state0[0][3 + s] = 1.0
```
Use the value of 3 as a Flag

ElnazDavoodi replied:
Use the value of 3 as a Flag

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_qlearner.py:
```diff
@@ -0,0 +1,297 @@
+            if centralized:
+                info_state0[0][3 + s] = 1.0
+            else:
+                info_state0[3 + s] = 1.0
```
Use the value of 3 as a Flag

ElnazDavoodi replied:
Use the value of 3 as a Flag

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_qlearner.py:
```diff
@@ -0,0 +1,297 @@
+                                             np.arange(num_messages), 0)
+            info_state1 = copy.deepcopy(base_info_state1)
+            if centralized:
+                info_state1[0][3 + s] = 1.0
```
Use the value of 3 as a Flag

ElnazDavoodi replied:
Use the value of 3 as a Flag

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_qlearner.py:
```diff
@@ -0,0 +1,297 @@
+            info_state1 = copy.deepcopy(base_info_state1)
+            if centralized:
+                info_state1[0][3 + s] = 1.0
+                info_state1[1][3 + m] = 1.0
```
Use the value of 3 as a Flag

ElnazDavoodi replied:
Use the value of 3 as a Flag

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_qlearner.py:
```diff
@@ -0,0 +1,297 @@
+                info_state1[0][3 + s] = 1.0
+                info_state1[1][3 + m] = 1.0
+            else:
+                info_state1[3 + m] = 1.0
```
Use the value of 3 as a Flag

ElnazDavoodi replied:
Use the value of 3 as a Flag

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_qlearner.py:
```diff
@@ -0,0 +1,297 @@
+
+def main(_):
+    game = "lewis_signaling"
+    num_players = 2
```
Use the value of 2 as a Flag

ElnazDavoodi replied:
Use the value of 2 as a Flag

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+
+def main(_):
+    game = "lewis_signaling"
+    num_players = 2
```
Use the value of 2 as a Flag

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+        payoffs = np.random.random((num_states, num_states))
+        payoffs_str = ",".join([str(x) for x in payoffs.flatten()])
+    elif FLAGS.payoffs == "climbing":
+        payoffs = np.array([[11, -30, 0], [-30, 7, 6], [0, 0, 5]]) / 30
```
Add comment about value of climbing payoffs for more readability

ElnazDavoodi replied:
Add comment about value of climbing payoffs for more readability

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_qlearner.py:
```diff
@@ -0,0 +1,297 @@
+        payoffs = np.random.random((num_states, num_states))
+        payoffs_str = ",".join([str(x) for x in payoffs.flatten()])
+    elif FLAGS.payoffs == "climbing":
+        payoffs = np.array([[11, -30, 0], [-30, 7, 6], [0, 0, 5]]) / 30
```
Add comment about value of climbing payoffs for more readability

ElnazDavoodi replied:
Add comment about value of climbing payoffs for more readability

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_qlearner.py:
```diff
@@ -0,0 +1,297 @@
+            if scalar_labels is None:
+                scalar_labels = [None] * len(scalars)
+
+            assert len(scalars) == len(
```
use assertEqual

vbhatt-cs replied:
Google code style recommended using exceptions for argument checking. I will convert the assert statements to exceptions unless assertEqual is better.

ElnazDavoodi commented on open_spiel/games/lewis_signaling.cc:
```diff
@@ -135,12 +135,16 @@ void LewisSignalingState::ObservationTensor(Player player,
-    (*values)[offset + message_] = 1;
-    offset += num_messages_;
+  if (static_cast<Players>(player) == Players::kSender) {
+    if (state_ != -1) {
```
only in the last step state_ == -1, is that right?

vbhatt-cs replied:
State is -1 before the chance node selects it. So it is just an additional check that the state has been assigned.

lanctot replied:
Hi @vbhatt-cs , can you change the -1 to `pyspiel.PlayerId.CHANCE` ?

lanctot replied:
Oh wait this is in C++, nevermind. Maybe just add a comment to explain this?

vbhatt-cs replied:
Would making a new const variable and doing `state_ != kUnassignedValue` work?

lanctot replied:
Yes, would be great (keep it in the Lewis signaling game namespace)

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+
+def main(_):
+    game = "lewis_signaling"
+    num_players = 2
```
Use the value of 2 as a Flag

vbhatt-cs replied:
Since the game always has only 2 players, I just used a variable and not a flag.

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+            for cur_episode in range(training_episodes):
+                time_step = env.reset()
+                # Find cur_state for logging. See lewis_signaling.cc for info_state details
+                cur_state = time_step.observations["info_state"][0][3:].index(1)
```
how about adding current state to time_step.observation["state"]?

vbhatt-cs replied:
I would have to change how rl_environment works. Currently it only stores "info_state", "legal_actions", "current_player" ([rl_environment.py](https://github.com/deepmind/open_spiel/blob/a651ce7fbea858b0be2db98afabb49f7c830d904/open_spiel/python/rl_environment.py#L219)). So, I thought it would be easier to just parse the info_state.

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+            base_info_state1 = [0.0, 1.0, 0.0] + [0.0] * num_states
+
+            for s in range(num_states):
+                info_state0 = copy.deepcopy(base_info_state0)
```
Not sure if you need a copy! why not passing base_info_state0 to agents[0]._epsilon_greedy?

vbhatt-cs replied:
I didn't find a way to artificially loop through all possible states and find the action given by the policy. This block of code was a hack to get it working. I modify the base_info_state to match the info_state that would be received by the agent and pass it to the action selection method. I used a similar method for the tabular case. If there is a better way to do it, I can refactor the complete block of code.

ElnazDavoodi commented on open_spiel/python/examples/lewis_signaling_dqn.py:
```diff
@@ -0,0 +1,255 @@
+
+            for s in range(num_states):
+                info_state0 = copy.deepcopy(base_info_state0)
+                info_state0[3 + s] = 1.0
```
use a value specified as a Flag instead of 3

vbhatt-cs replied:
3 corresponds to the initial indices of the info_state that doesn't change. I can convert it into a variable. If there is a better way to loop through the states, then the whole thing can be written in a better way.

lanctot commented:
Hi @vbhatt-cs , just wondering if you still plan to make the requested changes?

vbhatt-cs commented:
I have made the below changes:

1. Changed argument value checks in functions from assert to exception
2. Used kUnassignedValue to denote unassigned state and message in Lewis signaling game
3. Added comments about climbing game payoffs
4. Cleaned up the flags

I still haven't changed the code to loop through the states to make the final policy matrix. I had written it as a hack to make the figures. If there is a better way to loop through the states and get the greedy action than what I have done, then all the related code issues can be fixed.

lanctot commented:
Sorry for the delay @vbhatt-cs !

@ElnazDavoodi I think this is currently ok. Can you let me know if you have any further requests for changes before I pull it in for import?
